File: services/calc/debrief_calc/tools/area_summary.py
"""
Area summary tool.

Summarises features within a geographic region. Accepts either:

- An explicit bounding box via ``context.bounds`` (REGION-style context), or
- A list of selected features whose coordinates are unioned into a bbox
  (MULTI-style context).

The TypeScript counterpart at
``apps/web-shell/src/tools/region/analysis/areaSummary.ts`` accepts the same
two inputs: ``params.bounds`` (mirrors ``context.bounds``) and the selected
features. Both implementations return ``[]`` when neither path yields a valid
bbox. See #107 (F-2.6) for the alignment record.
"""

from __future__ import annotations

import math
import uuid
from typing import Any

from debrief_calc.models import (
    ContextType,
    GeoJSONFeatureDict,
    SelectionContext,
    ToolCategoryEnum,
    ToolParameter,
)
from debrief_calc.registry import tool
from debrief_schemas import OutputKindEnum


def _point_in_bbox(lon: float, lat: float, bbox: list[float]) -> bool:
    """Check if a point is within a bounding box [minx, miny, maxx, maxy]."""
    minx, miny, maxx, maxy = bbox
    return minx <= lon <= maxx and miny <= lat <= maxy


def _calculate_bbox_area_sq_nm(bbox: list[float]) -> float:
    """
    Calculate approximate area of a bounding box in square nautical miles.
    Uses a simple rectangular approximation.
    """
    minx, miny, maxx, maxy = bbox
    width_deg = maxx - minx
    height_deg = maxy - miny
    avg_lat = (miny + maxy) / 2
    height_nm = height_deg * 60
    width_nm = width_deg * 60 * math.cos(math.radians(avg_lat))
    return width_nm * height_nm


def _flatten_coords(coords: list) -> list[list[float]]:  # type: ignore[type-arg]
    """Recursively extract all [lon, lat] positions from GeoJSON coordinates."""
    if not isinstance(coords, list):
        return []
    if len(coords) > 0 and isinstance(coords[0], (int, float)):
        return [coords]
    result: list[list[float]] = []
    for item in coords:
        result.extend(_flatten_coords(item))
    return result


def _bounds_from_features(features: list[GeoJSONFeatureDict]) -> list[float] | None:
    """Extract bounding box from feature coordinates (matching TS approach)."""
    min_lon = float("inf")
    min_lat = float("inf")
    max_lon = float("-inf")
    max_lat = float("-inf")

    for f in features:
        geom = f.get("geometry")
        if not geom or "coordinates" not in geom:
            continue
        for pos in _flatten_coords(geom["coordinates"]):
            if len(pos) >= 2:
                lon, lat = pos[0], pos[1]
                if lon < min_lon:
                    min_lon = lon
                if lon > max_lon:
                    max_lon = lon
                if lat < min_lat:
                    min_lat = lat
                if lat > max_lat:
                    max_lat = lat

    if not math.isfinite(min_lon):
        return None
    return [min_lon, min_lat, max_lon, max_lat]


@tool(
    name="area-summary",
    description="Summarize the geographic extent and properties of selected features or a region",
    # Canonical FeatureKindEnum values that yield a meaningful bbox.
    # Kept in sync with apps/web-shell/src/tools/region/analysis/areaSummary.ts (#107).
    input_kinds=["TRACK", "POINT", "RECTANGLE", "CIRCLE", "ZONE", "POLY"],
    output_kind=OutputKindEnum.regionSOLIDUSstatistics,
    context_type=ContextType.MULTI,
    category=ToolCategoryEnum.calc,
    parameters=[
        ToolParameter(
            name="include_centroid",
            type="boolean",
            description="Include centroid point in output",
            default=True,
        )
    ],
)
def area_summary(context: SelectionContext, params: dict[str, Any]) -> list[GeoJSONFeatureDict]:
    """
    Summarize a geographic region.

    Accepts both explicit bounds (REGION context) and feature coordinates
    (MULTI context), aligning Python and TypeScript implementations (#107).

    Args:
        context: SelectionContext with bounds or features
        params: Optional parameters (include_centroid)

    Returns:
        List containing one Feature with area summary statistics
    """
    include_centroid = params.get("include_centroid", True)

    # Try explicit bounds first (REGION context), fall back to feature coordinates
    bounds = context.bounds
    if not bounds or len(bounds) != 4:
        bounds = _bounds_from_features(context.features)
    if not bounds or len(bounds) != 4:
        return []

    minx, miny, maxx, maxy = bounds

    # Calculate area
    area_sq_nm = _calculate_bbox_area_sq_nm(bounds)

    # Calculate dimensions
    import math

    avg_lat = (miny + maxy) / 2
    width_nm = (maxx - minx) * 60 * math.cos(math.radians(avg_lat))
    height_nm = (maxy - miny) * 60

    statistics: dict[str, Any] = {
        "area_sq_nm": round(area_sq_nm, 2),
        "width_nm": round(width_nm, 2),
        "height_nm": round(height_nm, 2),
    }

    if include_centroid:
        centroid_lon = (minx + maxx) / 2
        centroid_lat = (miny + maxy) / 2
        statistics["centroid"] = [round(centroid_lon, 4), round(centroid_lat, 4)]

    result_feature = {
        "type": "Feature",
        "id": f"area-{uuid.uuid4().hex[:8]}",
        "properties": {
            "statistics": statistics,
            "bounds": bounds,
        },
        "geometry": {
            "type": "Polygon",
            "coordinates": [[[minx, miny], [maxx, miny], [maxx, maxy], [minx, maxy], [minx, miny]]],
        },
    }

    return [result_feature]

File: apps/web-shell/src/tools/region/analysis/areaSummary.ts
/**
 * Area Summary tool — TypeScript implementation.
 *
 * Summarises the geographic extent and area of either an explicit bounding
 * box (`params.bounds`) or the coordinates of the selected features.
 * Aligned with the Python `area_summary` tool (#107).
 */

import type { DebriefFeature } from '@debrief/schemas';
import { OutputKindEnum } from '@debrief/schemas';
import type { MCPToolDefinition } from '@debrief/utils';
import type { RawGeoJSONFeature as GeoJSONFeature } from '@debrief/schemas';

type Position = number[];

function generateUUID(): string {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) return crypto.randomUUID();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export const toolDefinition: MCPToolDefinition = {
  name: 'area-summary',
  description:
    'Summarise the geographic extent of selected features, ' +
    'producing a bounding-box polygon with area statistics.',
  inputSchema: {
    type: 'object',
    properties: {
      features: { type: 'array', items: { type: 'object' } },
      params: {
        type: 'object',
        properties: {
          include_centroid: {
            type: 'boolean',
            default: true,
            description: 'Include centroid point in output',
          },
          bounds: {
            type: 'array',
            items: { type: 'number' },
            minItems: 4,
            maxItems: 4,
            description: 'Explicit bounding box [minx, miny, maxx, maxy]. ' +
              'When provided, takes precedence over feature-coordinate extraction.',
          },
        },
      },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [
      { kind: 'TRACK', min: 1 },
      { kind: 'POINT', min: 1 },
      { kind: 'RECTANGLE', min: 1 },
      { kind: 'CIRCLE', min: 1 },
      { kind: 'ZONE', min: 1 },
      { kind: 'POLY', min: 1 },
    ],
    'debrief:category': 'region/analysis',
    'debrief:version': '1.0.0',
    'debrief:outputKind': OutputKindEnum.regionSOLIDUSstatistics,
  },
};

/** Recursively extract all [lon, lat] positions from any GeoJSON coordinates. */
function flattenCoords(coords: unknown): Position[] {
  if (!Array.isArray(coords)) return [];
  if (typeof coords[0] === 'number') return [coords as Position];
  return coords.flatMap(flattenCoords);
}

/** Validate a 4-element numeric bounding box, returning it normalised or null. */
function parseBoundsParam(raw: unknown): [number, number, number, number] | null {
  if (!Array.isArray(raw) || raw.length !== 4) return null;
  const nums = raw.map(Number);
  if (nums.some((n) => !Number.isFinite(n))) return null;
  return [nums[0], nums[1], nums[2], nums[3]];
}

/** Extract a bounding box from feature coordinates, or null when none have valid coords. */
function boundsFromFeatures(
  features: DebriefFeature[],
): [number, number, number, number] | null {
  let minLon = Infinity, minLat = Infinity;
  let maxLon = -Infinity, maxLat = -Infinity;

  for (const f of features) {
    if (!f.geometry || !('coordinates' in f.geometry)) continue;
    for (const [lon, lat] of flattenCoords(f.geometry.coordinates)) {
      if (lon < minLon) minLon = lon;
      if (lon > maxLon) maxLon = lon;
      if (lat < minLat) minLat = lat;
      if (lat > maxLat) maxLat = lat;
    }
  }

  if (!isFinite(minLon)) return null;
  return [minLon, minLat, maxLon, maxLat];
}

export function execute(
  features: DebriefFeature[],
  params: Record<string, unknown>,
): GeoJSONFeature[] {
  const includeCentroid = params?.include_centroid !== false;

  // Try explicit bounds first (mirrors Python REGION-context behaviour),
  // then fall back to feature-coordinate extraction. Returns [] if neither
  // yields a valid bbox — aligning with the Python implementation (#107).
  const bounds = parseBoundsParam(params?.bounds) ?? boundsFromFeatures(features);
  if (!bounds) return [];

  const [minLon, minLat, maxLon, maxLat] = bounds;

  // Area calculation with latitude correction
  const avgLat = (minLat + maxLat) / 2;
  const widthDeg = maxLon - minLon;
  const heightDeg = maxLat - minLat;
  const widthNm = widthDeg * 60 * Math.cos((avgLat * Math.PI) / 180);
  const heightNm = heightDeg * 60;
  const areaSqNm = widthNm * heightNm;

  const statistics: Record<string, unknown> = {
    area_sq_nm: Math.round(areaSqNm * 100) / 100,
    width_nm: Math.round(widthNm * 100) / 100,
    height_nm: Math.round(heightNm * 100) / 100,
  };

  if (includeCentroid) {
    statistics.centroid = [
      Math.round((minLon + maxLon) / 2 * 10000) / 10000,
      Math.round((minLat + maxLat) / 2 * 10000) / 10000,
    ];
  }

  // Build a DatasetEnvelope for tabular display (#177)
  const tableDataset = {
    type: 'region_statistics',
    title: 'Area Summary',
    displayHint: 'table' as const,
    metadata: {
      xAxis: { label: 'Metric', type: 'nominal' as const },
      yAxis: { label: 'Value', type: 'quantitative' as const },
    },
    data: Object.entries(statistics)
      .filter(([, v]) => typeof v === 'number')
      .map(([key, val]) => ({
        metric: key.replace(/_/g, ' '),
        value: val,
      })),
  };

  return [{
    type: 'Feature',
    id: `area-${generateUUID()}`,
    geometry: {
      type: 'Polygon',
      coordinates: [[
        [minLon, minLat],
        [maxLon, minLat],
        [maxLon, maxLat],
        [minLon, maxLat],
        [minLon, minLat],
      ]],
    },
    properties: {
      kind: toolDefinition.annotations['debrief:outputKind'],
      name: `Area Summary`,
      label: `Area Summary`,
      statistics,
      bounds: [minLon, minLat, maxLon, maxLat],
      '__datasets': [tableDataset],
      style: {
        fill: true,
        fill_color: '#FF9800',
        fill_opacity: 0.1,
        stroke: true,
        color: '#FF9800',
        weight: 2,
        opacity: 0.8,
        dash_array: '5, 5',
      },
    },
  }];
}
